crates/bevy_ecs/src/observer/entity_observer.rs

```diff
@@ -0,0 +1,40 @@
+use super::*;
+use crate::component::{ComponentHooks, StorageType};
+
+/// Command to attach an entity observer to an entity
+pub(crate) struct AttachObserver {
+    pub(crate) target: Entity,
+    pub(crate) observer: Entity,
+}
+
+impl Command for AttachObserver {
+    fn apply(self, world: &mut World) {
+        if let Some(mut target) = world.get_entity_mut(self.target) {
+            let mut observed_by = target
+                .entry::<ObservedBy>()
+                .or_insert_with(|| ObservedBy(vec![]));
+            observed_by.0.push(self.observer);
+        } else {
+            world.despawn(self.observer);
+        }
+    }
+}
+
+/// Tracks a list of entity observers for the attached entity
+pub(crate) struct ObservedBy(Vec<Entity>);
```

Comment:
Alrighty here is another design sketch, this time for a simplified Observers as Separate Entities approach:

* Make the observer component/entity directly spawn-able. This makes the data model much clearer (and directly accessible) to users. It also removes the need for special "observer syntax" in the scene system.
* Do Observer system init inside the on_add component hook for the Observer component. This both (1) allows users to directly define a system on the observer component and (2) removes the need for manual registration of observer system types (cutting down on the boilerplate we see in the `observers.rs` example and unlocking the ability to add the `app.add_observer` methods as discussed)
* Make ObservedBy scoped to a specific observer type and use it as the source of truth for observers (now called `Observers<T>`)
* Note that I've omitted the optional `B: Bundle` type in my sketch for simplicity, but we can add that back in.
* Moved all state for a given observer into a single Observer component
* This does notably fully embrace Rust typing to achieve simplicity of dataflow (removes the scattering of sometimes redundant state across world internals and components on entities). This is intentional. The system in the PR does not currently support dynamic components either, and it would require significant additions to do so. We could create a dynamic component approach parallel (but similar to) to the approach defined in my sketch, and we would probably come out on top when it comes to both code size and code complexity.
* For this approach, imo it makes since to call the "observer component" Observer, as it represents the "observer object". I think we should consider renaming the current `Observer` type passed into the observer function to something else. I think `Trigger` makes more sense (even in the context of the current state of this PR), as that is really the scope of the data we are passing in  ... we aren't really passing in an Observer. This would then clash with the Trigger trait name ... we can discuss the best way to resolve this.

This is what it looks like from a user perspective (although we could / probably should still add helper methods for things like EntityMut):

```rust
// No manual init of types necessary
world.spawn(
    Observer::new(
        |explode: Trigger<Explode>, query: Query<&Mine>, mut commands: Commands| {
            // Do observer stuff here
        }
    )
    .target(id),
);

// later
commands.add(EntityTrigger::new(Explode, e));
```

And here is the code that backs it (standalone from the observer infra in this PR, other than the ObserverSystem stuff):

```rust
/// Tracks a list of entity observers for the attached entity
pub(crate) struct Observers<T> {
    entities: Vec<Entity>,
    _marker: PhantomData<T>,
}

impl<T> Default for Observers<T> {
    fn default() -> Self {
        Self {
            entities: Default::default(),
            _marker: Default::default(),
        }
    }
}

impl<T: Send + Sync + 'static> Component for Observers<T> {
    const STORAGE_TYPE: StorageType = StorageType::SparseSet;

    fn register_component_hooks(hooks: &mut ComponentHooks) {
        hooks.on_remove(|mut world, entity, _| {
            let mut component = world.get_mut::<Self>(entity).unwrap();
            let observed_by = std::mem::take(&mut component.entities);
            observed_by.iter().for_each(|&e| {
                if let Some(mut entity) = world.commands().get_entity(e) {
                    entity.despawn();
                };
            });
        });
    }
}

/// TODO: this being public does raise issues around "accessing observer system state inside of an observer system"
/// We need a way to protect against this, either via privacy or via
pub struct Observer<T: 'static> {
    system: Box<dyn ObserverSystem<T, ()>>,
    targets: Vec<Entity>,
}

impl<T: 'static> Observer<T> {
    pub fn new<M>(system: impl IntoObserverSystem<T, (), M>) -> Self {
        Self {
            system: Box::new(IntoObserverSystem::into_system(system)),
            targets: Vec::new(),
        }
    }

    pub fn target(mut self, entity: Entity) -> Self {
        self.targets.push(entity);
        self
    }
}

impl<T: Send + Sync + 'static> Component for Observer<T> {
    const STORAGE_TYPE: StorageType = StorageType::Table;

    fn register_component_hooks(hooks: &mut ComponentHooks) {
        hooks.on_add(|mut world, entity, _| {
            world.commands().add(move |world: &mut World| {
                // Track Targets
                // PERF: use get_entity_unchecked here and check for overlap inline to avoid the need to clone
                let targets = if let Some(entity_ref) = world.get_entity(entity) {
                    entity_ref.get::<Self>().unwrap().targets.clone()
                } else {
                    Vec::new()
                };

                let mut found = false;
                for target in targets {
                    if let Some(mut target) = world.get_entity_mut(target) {
                        let mut observers = target
                            .entry::<Observers<T>>()
                            .or_insert_with(Default::default);
                        observers.entities.push(entity);
                        found = true;
                    } else {
                    }
                }

                // clean up observer if there are no targets
                if !found {
                    world.despawn(entity);
                }

                // Initialize System
                let system: *mut dyn ObserverSystem<T, ()> =
                    if let Some(mut observe) = world.get_mut::<Self>(entity) {
                        &mut *observe.system
                    } else {
                        return;
                    };
                // SAFETY: World reference is exclusive and initialize does not touch system, so references do not alias
                unsafe {
                    (*system).initialize(world);
                }
            });
        });
    }
}

pub struct EntityTrigger<T> {
    trigger: T,
    entity: Entity,
}

impl<T> EntityTrigger<T> {
    pub fn new(trigger: T, entity: Entity) -> Self {
        Self { trigger, entity }
    }
}

impl<T: Send + Sync + 'static> Command for EntityTrigger<T> {
    fn apply(mut self, world: &mut World) {
        let world = world.as_unsafe_world_cell();
        if let Some(entity) = world.get_entity(self.entity) {
            // SAFETY: no active entity / component accesses
            let mut entity = unsafe { EntityMut::new(entity) };
            if let Some(observers) = entity.get_mut::<Observers<T>>() {
                for observer_entity in &observers.entities {
                    // A trigger target should not be an observer
                    if *observer_entity == self.entity {
                        continue;
                    }
                    let Some(observer_entity_cell) = world.get_entity(*observer_entity) else {
                        continue;
                    };
                    // SAFETY: we have ensured that observer_entity != self.entity, so there is no aliasing
                    let mut observer_entity_mut = unsafe { EntityMut::new(observer_entity_cell) };
                    // TODO: How do we prevent accessing this from within the observer system
                    if let Some(mut observe) = observer_entity_mut.get_mut::<Observer<T>>() {
                        // TODO: this static lifetime seems like an arbitrary limitation introduced by ObserverSystem.
                        // Can we lift it?
                        // This static issue is also a problem in the PR's current impl
                        let trigger_data: &'static mut T =
                            unsafe { std::mem::transmute(&mut self.trigger) };
                        let trigger = Trigger::new(
                            trigger_data,
                            ObserverTrigger {
                                observer: *observer_entity,
                                source: self.entity,
                                trigger: ComponentId::new(0),
                            },
                        );
                        observe.system.update_archetype_component_access(world);
                        unsafe { observe.system.run_unsafe(trigger, world) };
                        unsafe {
                            observe.system.queue_deferred(world.into_deferred());
                        };
                    }
                }
            }
        }
    }
}
```

Reply:
This does omit "component observers" as well, but if we move to "components as entities", then I believe we can directly use the code above as written?

Reply:
No, this misunderstands component observers. An OnAdd trigger has 3 relevant IDs, the entity, the component id of the  component being added and the id of OnAdd itself. An observer may (and often does) want to listen to all times a given component is added to an entity, you can't model that by observing the component as an entity because then you have no way to know what entity it was added to.

The current implementation does support dynamic components, the builder allows you to add arbitrary component IDs, this is key for the query cache use case where the types are erased.

Reply:
I did _also_ omit "entity-less" / global observers, which would then store the typed list of observers on a resource instead of a target entity.

I also haven't figured in the "fast path" observer stuff yet. But I think this approach lends itself to the same type of special casing. We _could_ store the collections directly on world in those cases (or we could just cache pointers to locations in the ECS storage).

Reply:
> "but if we move to "components as entities", then I believe we can directly use the code above as written?"
> No, this misunderstands component observers.

Ah riiiiiight sorry my brain is fried :)
This would require additional casing. However I do think this would still lend itself to storing an `Observers<OnAdd>` collection on the component entity for the target component?

Reply:
Eh wait no. I'm going to stop typing for tonight. I'm reasonably convinced that this general approach is compatible with OnAdd + SomeComponent + Entity, but my brain is done for the day :)

Reply:
Yes but you still need to be able to target an entity and a set of components on it as well as differentiate between observers listening to triggers on the entity an entity vs the entity as a component.

Reply:
Couldn't making observers a part of the spawn API just be done by adding a new generic component with a hook?
```rs
pub struct Observer<T: Trigger, F: FnMut(..)> {
    sys: Option<F>,
}

impl<T: Trigger, F: FnMut(..) + Send + Sync + 'static> Component for Observer<T> {
    const STORAGE_TYPE: StorageType = StorageType::Table;

    fn register_component_hooks(hooks: &mut ComponentHooks) {
        hooks.on_add(|mut world, entity, _| {
            if let Some(observer) = world.entity(entity).get_mut::<Self>() {
                let observer = std::mem::take(&mut observer.sys);
                world.init_trigger::<T>();
                world.commands().add(move |world: &mut World| {
                    world.entity(entity).observe(observer);
                });
            }
            world.commands().entity(entity).remove::<Self>();
        });
    }
}
```